src/box/lua/config/applier/roles.lua

```diff
@@ -180,8 +242,16 @@ local function post_apply(config)
         end
     end
 
-    last_loaded = loaded
-    last_roles_ordered = roles_ordered
+    roles_state.last_loaded = loaded
+    roles_state.last_roles_ordered = roles_ordered
+    roles_state.last_config = table.deepcopy(config)
+
+    -- box.status watcher values should be already set, because the watcher
+    -- has been registered during apply.
+    assert(roles_state.last_box_status_value ~= nil)
+
+    -- Call on_box_status callbacks after the config is fully applied.
+    call_on_box_status_callbacks(config, roles_state.last_box_status_value)
```

Comment:
I would think about guarantees behind this API that we offer to a user.

Let's assume that we have roles `foo`, `bar`, `baz` placed in the configuration is this order.

We have the following order of initialization (loading) the roles: `foo`, `bar`, `baz`.

Let's assume (for simplicity) that the roles are independent or dependent as `baz` depends on `bar`, `bar` depends on `foo`. So, the reordering gives the same order: `foo`, `bar`, `baz`.

We have the following order of callbacks:

* `foo.validate`
* `bar.validate`
* `baz.validate`
* `foo.apply`
* `bar.apply`
* `baz.apply`

It is guaranteed that the first one ends its execution before the next is called. Also, it is guaranteed that all of them are executed as part of the configuration process (and we can't reach `ready` or `check_warnings` status until all of them are finished).

The `on_event` is different: it occurs on some arbitrary point of time, out of the configuration process.

However, we guarantee the order of calling them and that the callbacks that are triggered by one event are executed one after the previous finished.

We also guarantee that neither of the `on_event` callbacks are called before `apply` for.. one given role?.. for all the roles?

At least, it gives us some predictable behaviour.

It is still not clear, how one roles should notify another one that it finishes its runtime/database initialization. Maybe some service registry or just a suggestion about proper using of globals. It seems like we should leave it out of scope here.

Anyway, I would like to see some description about guarantees behind this API, so a user wouldn't feel itself on a vague ground. Can you verify my wording, adjust/rephrase/shorten if needed and include it into the documentation request?

Reply:
Another point for the documentation is how the errors are handled. Unlike `apply`, an error from the `on_event` callback is ignored (with logging a message on the `error` level). Even if the callback is called because the configuration is changed, its error is just logged and doesn't fail the (re)configuration process.

Next, it would be good to make it clear, whether the `on_event` callbacks are executed out of the (re)configuration process in terms of time: does it waiti for finishing them. IOW, whether the configuration is successfully applied only after execution of these callbacks (if called due to the configuration change) or it is performed in background just like as when the callbacks are called due to a `box.status` event.

I don't know, which behavior is more expected, TBH. It is OK to perform some long-running activity from a usual watcher callback? I guess, no. So, let's keep the synchronous execution as part of the (re)configuration process.

(In some future we need alerts about long execution of role's function or application initialization code. An internal applier execution.)

Reply:
Added a more detailed doc, describing the order of callback execution as well as what happens if a callback raises an error.